### ISSUE TITLE
Disks with multiple contents, easy method ?

### ISSUE THREAD

**Issue by MajorWilly** [libretro/vice-libretro#164]
Hi, I've been getting seriously into the C64 lately and noticed that a lot of disks have multiple things on like docs, or the second part of a game. Would there be a way to implement an easy method of running a selected file from the disk contents like regular Vice can (just double click the required file in the contents list in the file requester). From what I understand at the moment you have to -

Reset the C64
Type `LOAD"$",8`
Type `LIST`
Move the cursor to the required file
Type `LOAD` (before the first `"`) and `,8,1` (after the second `"`)
Press Return.
Type `RUN`
Press Return.

This is quite a marathon with the virtual keyboard ! :smile: 

Or maybe there is a free basic file manager that allows you to select and run files that can be baked into the Vice core ? You could then add to the 'Reset Type' core option, so when you press the Reset button on the Virtual Keyboard, it would show the disk contents and let you select the required file.

**Comment by sonninnos** [comment]
Won't be a simple task at all to make the core handle any of that. A cartridge version of FB64 would be killer.

At the moment I ease this with Action Replay MK4, which makes shortcuts for those commands. F1 does loading (* on empty row), F3 gets the list, F5 does LIST, F7 does RUN.

Those plus cursors on the left analog stick is good enough for me. I also have bonus shortcuts for those function keys on Start+D-Pad, but they needs external tinkering with AutoHotkey.

**Comment by sonninnos** [comment]
Found something usable: https://csdb.dk/release/index.php?id=162390

And to use the cart all you need to do is create `vicerc` in `system/vice/` like so:

´´´
[C64SC]
CartridgeFile="\<path to fibr.crt\>"
´´´

And remove the "SC" if not using x64sc, naturally.

**Comment by MajorWilly** [comment]
That's exactly the sort of thing that I was after ! Some games didn't list the contents though, it just showed a list of the Retroarch directory ?! And if True Drive Emulation is on, it seems to take even longer than the original. I didn't realize cartridges could be used on this. I might try you Action Replay method.

**Comment by MajorWilly** [comment]
The Action Replay method seems better ! All that's needed is a Reset, F3 to list the contents, move to the desired file, then press F1 to load it! Plus it has a FastLoader so TrueDrive can be left on. I found a site that has crt's up to Action Replay 6.0. I'm wondering what the best version would be best to use. The latest ?

**Comment by sonninnos** [comment]
The only problem with Action Replay is that some games simply do not work with it, even as PRG. And True Drive Emulation has to be on and D81s won't work with the cart at all.

So to destroy the need of hassling with `vicerc` on a game/format basis the core really needs the option to skip the cart. It shall go to the TODO list then.

**Comment by sonninnos** [comment]
Well if you like the direct autostart to work then the MK4 is the last version to use. All others go to the menu which require a button press to get the ready-prompt.

**Comment by MajorWilly** [comment]
I'll stick with MK4 then. Thanks for your help and the heads up that cartridges work !

I tried a couple of the games you listed -

3D Pinball - Pinball Power
Planet Golf

They both worked fine. I'm using the GB64 versions. Also, I tried 'Dragon Ninja' (EU version). When it asked me if I wanted to enable fastloader I said No but it reset to Basic. I tried again and said Yes to enable Fastloader and it worked !

I'll leave this open for a bit.

**Comment by MajorWilly** [comment]
@rsn8887 Do you or anyone else know of a cartridge/hack that can easily display the disk contents and allow you to run a different file on the disk with minimal effort ? It seems Action Replay MK4 could be the best bet as stated above but the Fastload may cause issues. So far I've only encountered one game that doesn't load with it (P0 Snake 64K) and may be related to a crash on Simpsons. Action Replay 6 Fastload seems much more compatible but the games don't autoload. 

So basically do you know of a CRT or program that could be incorporated into lr-vice to easily display the disk contents and run a different file (like standalone Vice can) ?

**Comment by rsn8887** [comment]
Back in the day I used Final Cartridge 3, it comes with a nice mouse interface where you control the pointer with joystick:
https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fupload.wikimedia.org%2Fwikipedia%2Fen%2Ff%2Ff1%2FTheFinalCartridge3.png&f=1&nofb=1

It doesn't work with all games though.

**Comment by MajorWilly** [comment]
That's a shame. I was hoping there was a way after all these years to let you list and select a file to load off disk with full compatibility and without having to type what I listed in the first post. I suppose the only option is Action Replay MK4 (press one button to list contents and another button to load the chosen file). Action Replay Fastload is supposed to be the best (after JiffyDOS) but reduces the compatibility.

**Comment by rsn8887** [comment]
It would be nice to have, but I am not sure why it is needed really. There are plenty of archives of games with one game per disk, where you can always `load"*",8,1`. So why not just use those?

**Comment by rsn8887** [comment]
There might be other options, I haven't kept up with C64 programs over all the years. Final Cartridge 3 was just a thing I remembered from my childhood.

**Comment by MajorWilly** [comment]
Some games have 2 or more game parts on the disk but the other parts have to be manually loaded. There are some games that are individually levelpacked for quickplay (like outrun EU version has 6 different routes) and some just have extras like alternate loading screens, docs etc. that are not linked from a menu. Action Replay MK4 is the only easy option so i'll close this then.

**Comment by rsn8887** [comment]
Yes I understand, I deleted my comment.

There might be other options, I haven't kept up with C64 programs over all the years. Final Cartridge 3 was just a thing I remembered from my childhood.

**Comment by Zombeaver** [comment]
I don't see any mention here of using .cmds, so I'll just throw this out there. That's how I handle starting a specific program among several on a disk. You can create a .cmd file that contains something like:

x64 -autostart "Disk.d64:program" (with "program" being replaced with the specific prg name)

and load it and it'll load that specific program on start. Granted, this doesn't help you if you don't know what the program's name is beforehand - there's a great program called [DirMaster](https://style64.org/dirmaster) that you can use to examine the contents of C64 disk images.
